### PR TITLE
[#10936] Show explicit enroll errors

### LINKED ISSUE #10936: Enrolling students: undocumented restriction on team name length
When an instructor enrolls students, it turns out the team length cannot be longer than 60 chars. However, the error message doesn't mention this limitation (but specifies lot of other rules).
image

Might apply to other fields as well.

As a side note, there is room to improve error messages in the enroll feature in other ways as well.

### PR DESCRIPTION
**Pull request by daongochieu2810**
Fixes #10936
 - Return error messages from backend through ```EnrollError``` class to be displayed by frontend.
 - Create a new column in the fail to enroll students table to show error messages.
 - Fix table layout so text will not overflow.
Screenshot from 2021-01-31 12-44-23

### CODE REVIEW

**Review by hcwong — COMMENTED** [review]
Any reason why the changes were made directly on the `StudentsData` response directly? 

I'm uncomfortable with the idea because `StudentsData` is used by many Actions. It seems weird to add a `failToEnrollStudents` field when this would only be used for the `EnrollStudentsAction`. 

An alternative is to create a custom response for this action, something like `EnrollStudentsData`. Do let me know what you think.

**Comment by daongochieu2810** [comment]
Yes I think having a separate response for this action is better

**Comment by hcwong** [comment]
> Yes I think having a separate response for this action is better

Okay, on that note since `EnrollError` does not extend `ApiOutput`, consider making it a nested static class. You can look at `SessionResultsData` for reference. Also consider renaming `EnrollError` since it can be misleading as it is not exactly an `Exception`.

File: src/main/java/teammates/ui/output/StudentsData.java [hunk 1]
```diff
@@ -28,4 +28,5 @@ public StudentsData(List<StudentAttributes> students) {
     public void setStudents(List<StudentData> students) {
         this.students = students;
     }
+
```
Comment (hcwong) [inline src/main/java/teammates/ui/output/StudentsData.java:31]: To delete i guess

File: src/main/java/teammates/ui/webapi/EnrollStudentsAction.java [hunk 2]
```diff
@@ -87,6 +89,8 @@ JsonResult execute() {
                 } catch (InvalidParametersException | EntityDoesNotExistException
                         | EntityAlreadyExistsException exception) {
                     // Unsuccessfully enrolled students will not be returned.
+                    failToEnrollStudents.add(new EnrollStudentsData.EnrollErrorResults(student.email,
+                            exception.getMessage()));
                     return;
```
Comment (hcwong) [inline src/main/java/teammates/ui/webapi/EnrollStudentsAction.java:94]: This return seems redundant

File: src/main/java/teammates/ui/webapi/EnrollStudentsAction.java [hunk 3]
```diff
@@ -96,10 +100,12 @@ JsonResult execute() {
                     enrolledStudents.add(newStudent);
                 } catch (InvalidParametersException | EntityAlreadyExistsException exception) {
                     // Unsuccessfully enrolled students will not be returned.
+                    failToEnrollStudents.add(new EnrollStudentsData.EnrollErrorResults(student.email,
+                            exception.getMessage()));
                     return;
```
Comment (hcwong) [inline src/main/java/teammates/ui/webapi/EnrollStudentsAction.java:105]: Same as above

File: src/main/java/teammates/ui/webapi/EnrollStudentsAction.java [hunk 4]
```diff
@@ -71,6 +72,7 @@ JsonResult execute() {
         Set<String> existingStudentsEmail =
                 existingStudents.stream().map(StudentAttributes::getEmail).collect(Collectors.toSet());
         List<StudentAttributes> enrolledStudents = new ArrayList<>();
+        List<EnrollStudentsData.EnrollErrorResults> failToEnrollStudents = new ArrayList<>();
```
Comment (hcwong) [inline src/main/java/teammates/ui/webapi/EnrollStudentsAction.java:75]: Nit: Feels weird to start a non-function variable name with a verb. Suggest making this a noun. Something like `enrolledUnsuccessfulStudents` (not the best example) or something better.

**Review by hcwong — APPROVED** [review]
Few comments, but nothing major. LGTM, just address the stuff before merging

**Comment by moziliar** [comment]
One comment here: is it good/possible for the instructor to fold the error messages and have expand all option? In and expanded view, should the # of failed enrolments exceed ~20, it'd be hard for the instructor to view all?

**Comment by damithc** [comment]
Can we make it easier for the instructor to trace back the error to the spreadsheet? e.g.,
  * give the matching row number in this table
  * mark the offending rows in some way e.g., turn the bg color to red

**Comment by damithc** [comment]
> One comment here: is it good/possible for the instructor to fold the error messages and have expand all option? In and expanded view, should the # of failed enrolments exceed ~20, it'd be hard for the instructor to view all?

hmm... given that the user is unlikely NOT to want to look at the errors, I think it is fine to show all errors by default. Otherwise it is just one more unnecessary step the user has to perform.

**Comment by madanalogy** [comment]
> Can we make it easier for the instructor to trace back the error to the spreadsheet? e.g.,
> 
> * give the matching row number in this table
> * mark the offending rows in some way e.g., turn the bg color to red

@daongochieu2810 any response to the comment above? If not something you feel is within the scope of this PR maybe you could help create a feature request issue for it

**Comment by daongochieu2810** [comment]
@madanalogy the highlighting is done in this PR #10968

**Review by madanalogy — APPROVED** [review]

**Pull request merged**